Refactor chart widget and update plugin versions; remove generated files, modified postgre to "work" with newer lib.

Universal-Invariant opened the pull request:
I installed the apk and it would not load in WSA.  So I forked and uploaded the packages, modified what was broken,and was able to get the app to install in WSA. There is an obvious issue that when scrolling down too much the screen goes black until scrolling up. 

Also nots ure if the connection code is correct or not as I hacked whatI could together just to get something to work.  I noticed, also, the audio voice plays pretty fast.

Universal-Invariant committed:
Refactor chart widget and update plugin versions; remove generated files, modified postgre to "work" with newer lib.

wuxialearn commented:
Thanks for opening this PR. Can you clarify if the APK worked for you or not on Android or was it only WSA that didn't work? Also, I'm surprised there are Postgres code changes required because the app doesn't actually use Postgres anymore. Are you trying to use Postgres with the app or was it just giving an error for some reason?

Universal-Invariant commented:
I did not try it on any real android device or android emulator. I've been looking for some app to learn Chinese that would give me the audio, pinyin, character + defs so I stumbled upon your app. I installed it in WSA and it simply would not get past the initial screen. I figured it was something to with WSA and so forked the project to see if I could debug the app to figure out why it was stalling. I went ahead and updated all the packages automatically as far as I could get it just to see if it would work.

The Postgres stuff wasn't working and looking at the docs it seemed they did a significant rewrite so I just hacked things a bit to see if I could get it to work or not. It took about an hour or so but I was able to fix. The other error was similar but I just returned text, at first, and was able to get the project to build. It took more time since I'm relatively new to git, vscode, dart, and flutter(I had to install it all).

I noticed, when messing with Flutter that it did mention something about windows. But I mainly just wanted to install on WSA since I typically use it a lot and since it was failing it kinda irked me ;) I've noticed a few apps have failed on WSA for various strange reasons and I thought I could find out why. 

I might try to build it for windows. It depends if I get some time to mess with it further. There are a few things I think that could be added options to change voice speed, maybe font scale. Probably something much easier for you to do.

To simplify: No, the postgres issues were from a file from the project. If it's not used at all then it might be best to remove it from the project. It is causing the PR to fail it seems. pg_update.dart was the main changes. The only other change was updating chart.dart. All the other changes were just minor stuff like ignoring generated files, getting  keystore to work, etc. Obviously you can see those changes in the diff.

I removed pg_update.dart and it's references in settings. The debug version works fine without it. The other changes is just dart fix. In release, on WSA, it's not showing everything. Not sure if it is on my end or what. Seems the screen is not being refreshed. Might the original problem and it doesn't happen in debug.

wuxialearn commented:
I'm still a bit confused what wasn't working and which change in the commit fixed it. There are many changes in your commit that seems unrelated to each other. For example, gradle changes, linux changes, postgres and flchart. I think this is due to many packages being updated at the same time. Probably only one of these changes fixes the WSA issue. Another possibility is that none of these changes fixed the issue, and that the fix is actually contained in the Flutter version you downloaded. Flutter recently starting using a new rendering engine written from scratch called Impeller for Android. I believe I enabled this rendering engine in the last release(?). The engine is new and may have contained bugs due to it expecting to be run on an Android device. Generally Vulkan drivers are better on PCs than Mobile devices, but perhaps the emulation layer was causing an issue somewhere. This assumes that the newer Flutter version, which contains a new Impeller version, did not have this bug. (According to https://github.com/wuxialearn/wuxialearn/blob/main/flutter-version.txt, the version on F-droid is using Flutter 3.271.) In any case, it's an interesting issue.

Also, you may already know this, but WSA has been [discontinued](https://learn.microsoft.com/en-us/windows/android/wsa/):

> Starting March 5, 2025, Windows Subsystem for Android™ and the Amazon Appstore are no longer available in the Microsoft Store. 

I'm not against supporting WSA, but I would like to understand what the actual issue is that is being fixed by this PR before merging it. I'm also making some large changes in the next day or so, so maybe it will fix naturally. I will leave this PR open for now as a tracking issue.

>There are a few things I think that could be added options to change voice speed, maybe font scale.

If you feel up to it, I'm glad to take a look at any PR for improvements. IIRC the issue with slowing the TTS speed was that it sounds more robotic the slower you run it. Some of the new neural TTS voices are a bit better in this regard. For now I've just been waiting to see if an accurate, fast, cross platform TTS solution will emerge that can run on mid range mobile devices. I did consider having a font size option at some point but I never got around to it.

Universal-Invariant commented:
The initial reasons was so that I could contribute to the project. I installed the app and it looked like it could serve my purposes. I like using WSA more than emulators as it is much faster to debug.

I haven't made any changes to the project except updating the packages using 'flutter pub upgrade' and this led to the problems with postgre and the chart.dart which is pretty basic. This was all to setup any future modifications I might make. (I wanted to make sure I could at least get the project built before doing anything)

Since the project doesn't actually use postgre as you said then simply removing should likely be done either way. Upgrading the packages doesn't seem to cause any problems as far as I can tell so maybe it should be done too? The other updates are just from dart fix which are insignificant changes and it seems to think it is useful. (it's just removing the explicit return type which seems to be redundant)

As far as TTS, ReadEra sounds quite good. The voices are typically quite natural slow or fast.  It seems to be google TTS but they sound much better. I don't know much about it but options you can select from many different versions of the same voice pack and most sound quite good. In any case it would be better to let the user decide the speed even if it has some issues.

It's ultimately up to you though.

wuxialearn commented:
Ok, thanks for explaining that. The pg_update.dart file was used to sync changes from https://github.com/wuxialearn/chinese_data_tool which was the UI tool I used to create the units and sentences. It handles checking that all words in a unit have at least one sentence using that word, for example, among other features like moving words to different units, changing the order of units etc. The reason I say it is not used anymore is that most of that work is done, but it could be used again in the future, so I don't want to remove it yet. As for the check failing, this is the error:

> Because every version of flutter_test from sdk depends on stream_channel 2.1.2 and postgres 3.5.4 depends on stream_channel ^2.1.4, flutter_test from sdk is incompatible with postgres 3.5.4.

What I believe is happening is that your Flutter version is probably 3.29, and that version seems to contain a new flutter_test  version which is not in the version GH actions is using (3.27). The dart pub tool automatically tries to give you the newest version that works with your package (see https://dart.dev/tools/pub/versioning) so when you ran pub upgrade it gave you that version, however that version would not work for 3.27. I do intend to update to 3.29 eventually. There are some database changes I want to make. However I'm not sure I want to upgrade the postgres version as the current one works already and I don't want to test that the new version works or not as I would need to set the connectioninfo file again and find my old postgres connection string etc.

Universal-Invariant committed:
Downgraded postgre to 3.4.8 as recommended by CI.

Universal-Invariant commented:
Since you know far more about this stuff than I do it may or may not be worth it to did through what I did. But in some ways I also did some of the work, if I did it right, to get it update. All I can say is that the debug version works in WSA fine. The release version runs on WSA but it seems to have some issue with sizing certain widgets which is likely WSA specific.

I reverted the 2nd commit that removed postgres to pass conflict resolution since I interpreted what you said as it wasn't needed. To be clear:

1. I simply used the flutter update tool to update all the versions. [I installed latest flutter for this project and went through it's steps about versioning issues. This created the issues with chart.dart and pg_update.dart]

2. I modified chart.dart, a very minor modification dealing how they refactored the call style. It's a one line change.

3. I modified pg_update.dart because the changes were significantly different. But this seemed to just be, mainly, refactorization issues. It is likely nothing breaking here or if it is it could be reverted relatively easily. 

4. I added some .gitignore stuff for the keystore properties.

All the other changes were due to the flutter pub update and dart update.

Like I said, I did the minimum amount of work to try to update packages to pass the build.  You could, if you wanted and had the time, essentially do the same: It likely would take less than 30 minutes. You likely would run into the same problems which really just exist in chart.dart and pg_update.dart. The latter being the more serious but relatively minor.  

I downgraded the postgres version to 3.4.8 per CI. Build works in debug. If you can check and see if the release works for you then it might be worth looking into merging it.  On my end, for example, the intro screen will show only intro and a small blue dash. If it's a WSA issue and not due to upgrading then I think there likely won't be any problems with the merge.

wuxialearn commented:
did you push the commit?

Universal-Invariant commented:
I guess not ;/ 

I installed the release apk on WSA and it too has problems. Similar but different. When it first opens it shows only the blue rect then clicking opens the main screen but then it doesn't respond correctly to inputs and stuff. Probably some strange issue with optimization or something. Once everything is settled and I get some time I'll see if I can figure it out. Could be something with flutter and WSA. Probably not a huge deal since it works on debug and, like you said, they are gonna discontinue WSA ;/  (I think WSA is awesome)

wuxialearn commented:
I see the commit now. I had ran `dart format` after this pull request, so now there are a lot merge conflict. Also the checks are not running at all now for some reason. Sorry about that, the codebase is usually not so hectic. I feel like I've wasted your time.

>Probably not a huge deal since it works on debug and, like you said, they are gonna discontinue WSA ;/ (I think WSA is awesome)

WSA sounded cool but I never tried it because I heard it was incompatible with WSL which I was already using. I can't find a source for that now though.

Universal-Invariant commented:
lol, well, I have no idea. It doesn't matter that much as the changes are pretty minimal. 

As far as WSL and WSA I am running them both without issue. WSA is worth installing if you are developing android stuff IMO because it's much faster for development. Although it can have some strange bugs at times. E.g., I was working on an app and somehow the file system was not being properly synced and so it would report incorrect file sizes and even load the files corrupted. The shared folder it uses must have some bug in it. But otherwise it works well.

wuxialearn commented:
Closing for now as stale. Feel free to re-open debased on current main if you'd like.
